gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobRewriter.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud.storage;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.gcloud.RetryHelper.runWithRetries;
+
+import com.google.api.services.storage.model.RewriteResponse;
+import com.google.gcloud.RetryHelper;
+import com.google.gcloud.spi.StorageRpc;
+
+import java.math.BigInteger;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+/**
+ * Google Storage blob rewriter.
+ */
+public final class BlobRewriter {
```

Comment:
We could make it `Restorable` and support re-write continuation between processes (similar to a write channel), no?